### PR TITLE
HDDS-5314. Show number of Open containers per Node in Recon UI

### PR DESCRIPTION
**Pull request by yuangu002**
## What changes were proposed in this pull request?

The Recon UI DataNode page only shows the total number of containers. This jira aims to show the open containers count on the UI.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-5314

## How was this patch tested?
Tested it on UI, and use "ozone admin container close" to close a container.

### CODE REVIEW

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java [hunk 1]
```diff
@@ -117,11 +123,26 @@ public Response getDatanodes() {
...
+        int containers = allContainers.size();
+        int closedContainers = containers - openContainers;
         builder.withContainers(containers);
+        builder.withOpenContainers(openContainers);
+        builder.withClosedContainers(closedContainers);
       } catch (NodeNotFoundException ex) {
         LOG.warn("Cannot get containers, datanode {} not found.",
             datanode.getUuid(), ex);
+      } catch (ContainerNotFoundException cnfe) {
+        LOG.warn("Cannot found container.", cnfe);
```
Comment (vivekratnavel) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java:145]: nit: 
```suggestion
        LOG.warn("Cannot find container.", cnfe);
```

**Review by vivekratnavel — COMMENTED** [review]
+1 overall LGTM.

Thanks for working on this! It would be good if you can also update `hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json` with openContainers and closedContainers. This is the json file used to mock a response from the server to run Recon UI locally. Also, it would help if you post a screen-shot of the UI after your change. Thanks!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/DatanodeMetadata.java [hunk 2]
```diff
@@ -56,6 +56,12 @@
   @XmlElement(name = "containers")
   private int containers;
 
+  @XmlElement(name = "openContainers")
+  private int openContainers;
+
+  @XmlElement(name = "closedContainers")
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/DatanodeMetadata.java:62]: Since we have other transient states from OPEN to CLOSED, maybe we can just track openContainers & totalContainers. It is the 'OPEN' container count per node which is a significant data point for writes.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java [hunk 3]
```diff
@@ -117,11 +123,26 @@ public Response getDatanodes() {
         }
       });
       try {
-        int containers = nodeManager.getContainers(datanode).size();
+        Set<ContainerID> allContainers = nodeManager.getContainers(datanode);
+
+        int openContainers = 0;
+        for (ContainerID containerID: allContainers) {
+          ContainerInfo containerInfo = reconContainerManager.getContainer(containerID);
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java]: When a container goes from OPEN to CLOSED it is meant to be an irreversible change, and it stays CLOSED (or DELETED).  Over the lifetime of the cluster, we expect the number of OPEN containers (which are newly created) to remain stable, while the CLOSED containers will keep increasing. If we iterate over all the containers in the system for every call, then most of the computation will remain unchanged (since CLOSED containers will remain that way).    Given the above, can we think of a way to optimize this using cached information?

**Review by avijayanhwx — COMMENTED** [review]
Thanks for working on this Yuan! A couple of comments inline.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/DatanodeMetadata.java (hunk 2 above)
Comment (yuangu002) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/DatanodeMetadata.java:62]: Got it, will fix that soon

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java (hunk 1 above)
Comment (yuangu002) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java]: Fixed.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java [hunk 4]
```diff
@@ -117,11 +123,26 @@ public Response getDatanodes() {
...
+          if (containerInfo.isOpen()) {
+            ++openContainers;
+          }
+        }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java]: Could this loop be eliminated? There might be an existing count lying somewhere.

File: hadoop-hdds/common/audit.log [hunk 5]
```diff
@@ -0,0 +1 @@
+ERROR | OMAudit | ? | user=john | ip=192.168.0.1 | op=CREATE_VOLUME {key1=value1, key2=value2} | ret=FAILURE
```
Comment (smengcl) [inline hadoop-hdds/common/audit.log:1]: pushed by accident?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java [hunk 6]
```diff
@@ -317,6 +317,8 @@ public void addContainer(final UUID uuid,
     lock.writeLock().lock();
     try {
       checkIfNodeExist(uuid);
+      // SCMContainerManager.getContainer() -> containerInfo
+
```
Comment (smengcl) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java:321]:
```suggestion
```

also accidental?

Comment (yuangu002) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java]: Yeah, demo lines. Fixed

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/CloseContainerEventHandler.java [hunk 7]
```diff
@@ -85,9 +85,11 @@ public void onMessage(ContainerID containerID, EventPublisher publisher) {
...
         command.setEncodedToken(getContainerToken(containerID));
 
-        getNodes(container).forEach(node ->
-            publisher.fireEvent(DATANODE_COMMAND,
-                new CommandForDatanode<>(node.getUuid(), command)));
+        getNodes(container).forEach(node -> {
+          publisher.fireEvent(DATANODE_COMMAND,
+                new CommandForDatanode<>(node.getUuid(), command));
+        });
+
```
Comment (smengcl) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/CloseContainerEventHandler.java:92]: Let's remove unnecessary changes in this file.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java [hunk 8]
```diff
@@ -108,6 +115,10 @@ public Response getDatanodes() {
           if (datanode.getUuid().equals(pipeline.getLeaderId())) {
             leaderCount.getAndIncrement();
           }
+          int openContainerPerPipeline =
+                  reconContainerManager.getPipelineToOpenContainer()
+                  .getOrDefault(pipelineID, 0);
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java:120]: What is the rationale behind defaulting to pipeline ID `0` here? Wouldn't `get(pipelineID)` be better?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java [hunk 9]
```diff
@@ -108,6 +115,10 @@ public Response getDatanodes() {
...
+          openContainers.set(openContainers.get() + openContainerPerPipeline);
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java:121]: Doesn't seem atomic to me.. Try this?
```suggestion
          openContainers.getAndAdd(openContainerPerPipeline);
```

**Review by smengcl — CHANGES_REQUESTED** [review]

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 10]
```diff
@@ -203,6 +207,13 @@ private void checkContainerStateAndUpdate(ContainerID containerID,
         && isHealthy(state)) {
       LOG.info("Container {} has state OPEN, but given state is {}.",
           containerID, state);
+      Pipeline pipeline =
+              pipelineManager.getPipeline(containerInfo.getPipelineID());
+      // subtract open container count from the pipeline
+      int curCnt = pipelineToOpenContainer.getOrDefault(pipeline.getId(), 0);
+      if (curCnt > 0) {
+        pipelineToOpenContainer.put(pipeline.getId(), curCnt - 1);
+      }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:216]:
```suggestion
      // We assume that this method is single-threaded
      int curCnt = pipelineToOpenContainer.getOrDefault(pipeline.getId(), 0);
      if (curCnt == 1) {
        pipelineToOpenContainer.remove(pipeline.getId());
      } else if (curCnt > 0) {
        pipelineToOpenContainer.put(pipeline.getId(), curCnt - 1);
      }
```

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 11]
```diff
@@ -223,12 +234,16 @@ public void addNewContainer(ContainerWithPipeline containerWithPipeline)
...
+        Pipeline pipeline = containerWithPipeline.getPipeline();
+        PipelineID pipelineID = pipeline.getId();
         if (pipelineManager.containsPipeline(pipelineID)) {
           getContainerStateManager().addContainer(containerInfo.getProtobuf());
           pipelineManager.addContainerToPipeline(
               containerWithPipeline.getPipeline().getId(),
               containerInfo.containerID());
+          // update open container count on all datanodes on this pipeline
+          pipelineToOpenContainer.put(pipelineID,
+                    pipelineToOpenContainer.getOrDefault(pipelineID, 0) + 1);
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:246]: Use AtomicInteger for the value

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 12]
```diff
@@ -70,6 +71,8 @@
   private final Table<UUID, DatanodeDetails> nodeDB;
   // Container ID -> Datanode UUID -> Timestamp
   private final Map<Long, Map<UUID, ContainerReplicaHistory>> replicaHistoryMap;
+  // Pipeline -> open containers
+  private Map<PipelineID, Integer> pipelineToOpenContainer;
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:75]:
```suggestion
  private Map<PipelineID, AtomicInteger> pipelineToOpenContainer;
```

Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:75]:
```suggestion
  private final Map<PipelineID, Integer> pipelineToOpenContainer;
```

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java (hunk 3 above)
Comment (yuangu002) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java]: fixed, please review.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 13]
```diff
@@ -70,6 +71,8 @@
   private final Table<UUID, DatanodeDetails> nodeDB;
   // Container ID -> Datanode UUID -> Timestamp
   private final Map<Long, Map<UUID, ContainerReplicaHistory>> replicaHistoryMap;
+  // Pipeline -> open containers
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:74]: nit
```suggestion
  // Pipeline -> # of open containers
```

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 14]
```diff
@@ -203,6 +207,15 @@ private void checkContainerStateAndUpdate(ContainerID containerID,
...
           containerID, state);
+      Pipeline pipeline =
+              pipelineManager.getPipeline(containerInfo.getPipelineID());
+      // subtract open container count from the pipeline
+      int curCnt = pipelineToOpenContainer.getOrDefault(pipeline.getId(), 0);
+      if (curCnt == 1) {
+        pipelineToOpenContainer.remove(pipeline.getId());
+      } else if (curCnt > 0) {
+        pipelineToOpenContainer.put(pipeline.getId(), curCnt - 1);
+      }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:218]: Can we use `containerInfo.getPipelineID()` directly?
Seems we are converting from `PipelineID` to `Pipeline` back to the same `PipelineID`?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 15]
```diff
@@ -223,12 +236,16 @@ public void addNewContainer(ContainerWithPipeline containerWithPipeline)
     ContainerInfo containerInfo = containerWithPipeline.getContainerInfo();
     try {
       if (containerInfo.getState().equals(HddsProtos.LifeCycleState.OPEN)) {
-        PipelineID pipelineID = containerWithPipeline.getPipeline().getId();
+        Pipeline pipeline = containerWithPipeline.getPipeline();
+        PipelineID pipelineID = pipeline.getId();
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:240]: Let's also restore these two lines, doesn't seem necessary any more.

**Review by smengcl — APPROVED** [review]
+1.

Let's add a unit test where we check this works when there are 10+ open containers (and preferrably 2+ pipelines). And then try to close containers one by one to see the open container count decrease accordingly.

**Review by vivekratnavel — APPROVED** [review]
+1 LGTM

**Comment by vivekratnavel** [comment]
@yuangu002 Please fix the checkstyle warnings.

**Comment by smengcl** [comment]
Thanks @yuangu002 for the contribution. And thanks @avijayanhwx @vivekratnavel for the review.
